### ISSUE TITLE
Nvidia DALI preprocessing

### ISSUE THREAD

**Issue by ordioma** [ultralytics/ultralytics#16460]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I use yolov8 for tracking and i need to reduce CPU load. Frames are received from a rstp stream one by one, so i the question is if Nvidia DALI can help and if so how to do it in python?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ordioma, thank you for your interest in Ultralytics 🚀! 

We recommend checking our [Docs](https://docs.ultralytics.com) for useful insights. If you're facing performance issues and considering using Nvidia DALI, it may be worth looking into our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to optimize your setup.

If this is a 🐛 Bug Report, make sure to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/).

For further support, join our community on [Discord](https://ultralytics.com/discord) 🎧 for real-time chat, participate in discussions on [Discourse](https://community.ultralytics.com), or explore our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you're using the latest `ultralytics` package. Upgrade with:

```bash
pip install -U ultralytics
```

Make sure your environment meets all requirements as detailed in our [repository](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml).

## Environments

YOLOv8 is compatible with the following environments, ensuring all dependencies are pre-installed:

- **Notebooks** (with free GPU): <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Platforms**: Check out our guides for [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) and [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Refer to our [Docker Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If the badge is green, all our [CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are successfully passing, ensuring the stability across different systems.

This is an automated response, but an Ultralytics engineer will be with you shortly to assist with your query. 😊

**Comment by Y-T-G** [comment]
You can use NVIDIA's hardware decoding to reduce CPU load. For that, you would need to use OpenCV with GStreamer pipeline and `nvuridecodebin` for video decoding. Typically, this is part of the DeepStream docker image. The normal OpenCV installation doesn't come with GStreamer support.

We also have a guide for DeepStream for Jetson.

https://docs.ultralytics.com/guides/deepstream-nvidia-jetson/#what-is-nvidia-deepstream

**Comment by ordioma** [comment]
> `nvuridecodebin`

I already use GPU video decoding
                            cap = func_timeout(10., lambda: cv2.VideoCapture(
                                f"rtspsrc location={url} latency=0 ! "
                                f"rtph264depay ! "
                                f"h264parse ! "
                                f"nvv4l2decoder ! "
                                f"nvvideoconvert ! "
                                f"videoconvert ! "
                                f"videorate ! "
                                f"video/x-raw,framerate={FPS}/1 ! "
                                f"appsink drop=1"))
So, I'm particularly interested in speeding up YOLO preprocessing and postprocessing.

**Comment by ordioma** [comment]
> For the preprocessing, you should keep the image on the GPU instead of moving it to the CPU with the `nvvideoconvert`. You can then use CV-CUDA for preprocessing.
> 
> https://github.com/CVCUDA/CV-CUDA/blob/main/DEVELOPER_GUIDE.md
> 
> After that, you can create a torch tensor directly from the GPU image and then pass it to the `predict` method. When you pass a torch tensor to `predict`, it assumes that the tensor has already been preprocessed so it will not perform preprocessing again and directly perform inference on it.

Could you provide minimal example how to do it?

**Comment by glenn-jocher** [comment]
Certainly! Here's a minimal example:

```python
import torch
from ultralytics import YOLO

# Assuming `gpu_image` is your preprocessed image on the GPU
# Convert it to a torch tensor
tensor_image = torch.from_numpy(gpu_image).to('cuda')

# Load the model
model = YOLO("yolov8n.pt")

# Run inference
results = model.predict(tensor_image)
```

This assumes you have the image already on the GPU. Adjust the `gpu_image` part to match your setup.

**Comment by ordioma** [comment]
I face the problem that with .track it doesn't work, but with . predict it works just fine
![image](https://github.com/user-attachments/assets/e35a21bb-6788-436d-b16d-212426a2baf1)
![image](https://github.com/user-attachments/assets/93fe7380-dacc-4c30-bf0c-51c5319c2de2)

**Comment by Y-T-G** [comment]
Try `tracker=bytesort.yaml`.

`botsort` uses optical flow. It's both slow and CPU intensive.

**Comment by wanjunling168** [comment]
> Try `tracker=bytesort.yaml`.
> 
> `botsort` uses optical flow. It's both slow and CPU intensive.

should be `bytetrack.yaml` :)
there is only `botsort` and `bytetrack` in the official repo

**Comment by glenn-jocher** [comment]
Thank you for pointing that out! Yes, it should be `bytetrack.yaml`.
